coderd/tailnet.go

```diff
@@ -0,0 +1,340 @@
+package coderd
+
+import (
+	"bufio"
+	"context"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"net/netip"
+	"net/url"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+	"tailscale.com/derp"
+	"tailscale.com/tailcfg"
+
+	"cdr.dev/slog"
+	"github.com/coder/coder/coderd/wsconncache"
+	"github.com/coder/coder/codersdk"
+	"github.com/coder/coder/site"
+	"github.com/coder/coder/tailnet"
+)
+
+var defaultTransport *http.Transport
+
+func init() {
+	var valid bool
+	defaultTransport, valid = http.DefaultTransport.(*http.Transport)
+	if !valid {
+		panic("dev error: default transport is the wrong type")
+	}
+}
+
+// TODO: ServerTailnet does not currently remove stale peers.
+
+// NewServerTailnet creates a new tailnet intended for use by coderd. It
+// automatically falls back to wsconncache if a legacy agent is encountered.
+func NewServerTailnet(
```

Comment:
Although maybe a bit weird, I'd argue we should put this in `workspaceagents.go` since `wsconncache` will be going away.